src/Omega_h_pool_kokkos.cpp

```diff
@@ -0,0 +1,286 @@
+  return freeSetBySize.size();
+}
+
+auto StaticKokkosPool::getRequiredChunks(size_t n, size_t bytesPerChunk)
+    -> size_t {
+  return (n / bytesPerChunk) + (n % bytesPerChunk ? 1 : 0);
+}
+
+StaticKokkosPool::~StaticKokkosPool() {
+  Kokkos::kokkos_free(pool);
+}
+
+auto KokkosPool::getChunkSize() const -> size_t { return chunkSize; }
+
+auto KokkosPool::getNumFreeFragments() const -> unsigned {
+  unsigned numFreeFragments = 0;
+
+  for (const auto& pool : pools) {
+    numFreeFragments += pool.getNumFreeFragments();
+  }
+
+  return numFreeFragments;
+}
+
+KokkosPool::KokkosPool(size_t bytesPerChunks) : chunkSize(bytesPerChunks) {}
+
+auto KokkosPool::allocate(size_t n) -> void* {
+  auto current = pools.begin();
+  size_t mostAmountOfChunks = 0;
+
+  while (current != pools.end()) {
+    void* ptr = current->allocate(n);
+    if (ptr) {
+      allocations[ptr] = current;
+      return ptr;
+    }
+
+    if (current->getNumChunks() > mostAmountOfChunks) {
+      mostAmountOfChunks = current->getNumChunks();
+    }
+
+    current++;
+  }
+
+  size_t requestedChunks = StaticKokkosPool::getRequiredChunks(n, chunkSize);
+  size_t amortizedChunkSize = std::max(mostAmountOfChunks * 2, requestedChunks);
+
+  try {
+    pools.emplace_back(amortizedChunkSize, chunkSize);
+  } catch (const std::runtime_error& e) {
+
+    std::cerr << "Amortization with " << amortizedChunkSize << " chunks failed." << std::endl;
+    std::cerr << e.what() << std::endl;
+
+    if (amortizedChunkSize <= requestedChunks) {
+      throw;
+    }
+
+    pools.emplace_back(requestedChunks, chunkSize);
+  }
+
+  void* ptr = pools.back().allocate(n);
+  assert(ptr != nullptr);
+  allocations[ptr] = std::prev(pools.end());
+
+  return ptr;
+}
+
+void KokkosPool::deallocate(void* data) {
+  try {
+    allocations.at(data)->deallocate(data);
+  } catch (const std::out_of_range& e) {
+    std::cerr << "Attempted to deallocate data that was not allocated by this pool." << std::endl;
+    throw;
+  }
+  allocations.erase(data);
+}
+
+auto KokkosPool::getNumAllocations() const -> unsigned {
+  return allocations.size();
+}
+
+auto KokkosPool::getNumFreeChunks() const -> unsigned {
+  unsigned numFreeChunks = 0;
+
+  for (const auto& pool : pools) {
+    numFreeChunks += pool.getNumFreeChunks();
+  }
+
+  return numFreeChunks;
+}
+
+auto KokkosPool::getNumAllocatedChunks() const -> unsigned {
+  unsigned numAllocatedChunks = 0;
+
+  for (const auto& pool : pools) {
+    numAllocatedChunks += pool.getNumAllocatedChunks();
+  }
+
+  return numAllocatedChunks;
+}
+
+auto KokkosPool::getNumChunks() const -> unsigned {
+  unsigned numChunks = 0;
+
+  for (const auto& pool : pools) {
+    numChunks += pool.getNumChunks();
+  }
+
+  return numChunks;
+}
+
+auto KokkosPool::getGlobalPool() -> KokkosPool& {
+  static Omega_h::KokkosPool s_pool { 1000 };
+
+  if (s_pool.getNumChunks() == 0) {
+    s_pool.pools.emplace_back(700'000, s_pool.chunkSize);
```

Comment:
It allocates 700k chunks; each chunk is 1Kb as declared on line 267. This allocates an initial pool of 700k*1KB = 700MB which was chosen because it was slightly more than the most amount of memory the 500k delta wing case uses (> 600MB)